Sources/Testing/Running/Runner.Plan.swift

```diff
@@ -282,3 +282,132 @@ extension Runner.Plan {
     _independentlyRunnableSteps(in: stepGraph)
   }
 }
+
+// MARK: - Snapshotting
+
+extension Runner.Plan {
+  /// A serializable snapshot of a ``Runner/Plan-swift.struct`` instance.
+  @_spi(ExperimentalSnapshotting)
+  public struct Snapshot: Sendable {
+    /// The graph of the steps in this runner plan.
+    private var _stepGraph: Graph<String, Step.Snapshot?> = .init(value: nil)
+
+    /// Initialize an instance of this type by snapshotting the specified plan.
+    ///
+    /// - Parameters:
+    ///   - plan: The original plan to snapshot.
+    public init(snapshotting plan: Runner.Plan) async {
+      await plan.stepGraph.forEach { keyPath, step in
+        let step: Runner.Plan.Step.Snapshot? = if let step {
+          await Step.Snapshot(snapshotting: step)
+        } else {
+          nil
+        }
+        _stepGraph.insertValue(step, at: keyPath)
+      }
+    }
+
+    /// The steps of this runner plan.
+    public var steps: some Collection<Step.Snapshot> {
+      _stepGraph.compactMap(\.value)
+    }
+  }
+}
+
+extension Runner.Plan.Snapshot: Codable {
+  /// The coding keys used for serializing a
+  /// ``Runner/Plan-swift.struct/Snapshot`` instance.
+  private enum _CodingKeys: CodingKey {
```

Comment:
Would it help if we made `Graph` conditionally `Codable`? It ought to be trivial to do so.

Reply:
Yeah I'm not opposed to that as a general enhancement, but for this usage in particular it felt reasonable for the serialized representation of this to just be a simpler, array-like data data structure. It feels a bit more conservative and safer for us to encode it that way, since it reserves our ability to evolve `Graph`'s implementation further without affecting serialization. So for now I plan to proceed as-is, but we can keep thinking about it.